File: native/cpp/src/policy.cpp
#include "luck/policy.hpp"
#include <algorithm>
#include <chrono>
#include <cmath>
#include <utility>

namespace luck {
namespace {

void add_finding(PolicyDecision& decision,std::string rule,std::string message,RiskLevel risk,PolicyAction action) {
    decision.findings.push_back({std::move(rule),std::move(message),risk,action});
    if (action==PolicyAction::deny) decision.action=PolicyAction::deny;
    else if (action==PolicyAction::warn && decision.action==PolicyAction::allow) decision.action=PolicyAction::warn;
}

bool domain_is_within(std::string_view domain,std::string_view suffix) {
    const auto normalized=normalized_domain(domain);
    const auto expected=normalized_domain(suffix);
    if (normalized==expected) return true;
    return normalized.size()>expected.size() && normalized.ends_with(expected) &&
           normalized[normalized.size()-expected.size()-1U]=='.';
}

bool contains_case_insensitive(std::string_view value,std::string_view needle) {
    return ascii_lower(value).find(ascii_lower(needle))!=std::string::npos;
}

bool begins_with_case_sensitive(std::string_view value,std::string_view prefix) {
    return value.size()>=prefix.size() && value.substr(0U,prefix.size())==prefix;
}

double current_unix_seconds() {
    return std::chrono::duration<double>(std::chrono::system_clock::now().time_since_epoch()).count();
}

}

std::size_t PolicyDecision::count(RiskLevel risk) const noexcept {
    return static_cast<std::size_t>(std::count_if(findings.begin(),findings.end(),[risk](const PolicyFinding& finding){ return finding.risk==risk; }));
}

CookiePolicy::CookiePolicy(CookiePolicyConfig config):config_(std::move(config)){}
const CookiePolicyConfig& CookiePolicy::config() const noexcept { return config_; }

PolicyDecision CookiePolicy::evaluate(const Cookie& cookie) const { return evaluate(cookie,current_unix_seconds()); }

PolicyDecision CookiePolicy::evaluate(const Cookie& cookie,double now_unix_seconds) const {
    PolicyDecision decision;
    if (cookie.name.empty()) add_finding(decision,"syntax.name.empty","Cookie name is empty.",RiskLevel::high,PolicyAction::deny);
    if (cookie.domain.empty()) add_finding(decision,"syntax.domain.empty","Cookie domain is empty.",RiskLevel::high,PolicyAction::deny);
    if (cookie.path.empty() || cookie.path.front()!='/') add_finding(decision,"syntax.path.invalid","Cookie path must begin with a slash.",RiskLevel::high,PolicyAction::deny);
    if (cookie.value.find_first_of("\r\n")!=std::string::npos) add_finding(decision,"syntax.value.line_break","Cookie value contains a line break.",RiskLevel::critical,PolicyAction::deny);
    if (cookie.estimated_bytes()>config_.maximum_cookie_bytes) {
        add_finding(decision,"size.maximum","Cookie exceeds the configured byte limit.",RiskLevel::high,PolicyAction::deny);
    }
    if (config_.require_secure && !cookie.secure) add_finding(decision,"security.secure.required","Policy requires Secure.",RiskLevel::high,PolicyAction::deny);
    else if (!cookie.secure) add_finding(decision,"security.secure.disabled","Cookie can be sent over an unencrypted connection.",RiskLevel::high,PolicyAction::warn);
    if (!cookie.http_only) add_finding(decision,"security.http_only.disabled","Cookie can be read by page scripts.",RiskLevel::medium,PolicyAction::warn);
    if (cookie.same_site==SameSite::none && !cookie.secure) add_finding(decision,"security.same_site_none_without_secure","SameSite=None requires Secure.",RiskLevel::high,PolicyAction::deny);
    if (config_.deny_same_site_none && cookie.same_site==SameSite::none) add_finding(decision,"policy.same_site_none","Policy denies SameSite=None.",RiskLevel::high,PolicyAction::deny);
    if (cookie.same_site==SameSite::unspecified) add_finding(decision,"security.same_site.unspecified","Browser defaults determine cross-site behavior.",RiskLevel::low,PolicyAction::warn);
    if (config_.deny_session_cookies && cookie.effective_session()) add_finding(decision,"policy.session","Policy denies session cookies.",RiskLevel::medium,PolicyAction::deny);
    if (config_.deny_empty_values && cookie.value.empty()) add_finding(decision,"policy.empty_value","Policy denies empty cookie values.",RiskLevel::medium,PolicyAction::deny);
    if (config_.deny_expired_cookies && cookie.expired_at(now_unix_seconds)) add_finding(decision,"policy.expired","Cookie has already expired.",RiskLevel::medium,PolicyAction::deny);
    if (config_.require_host_only && !cookie.host_only) add_finding(decision,"policy.host_only","Policy requires host-only cookies.",RiskLevel::high,PolicyAction::deny);
    if (cookie.partitioned && !cookie.secure) add_finding(decision,"security.partitioned.secure","Partitioned cookies must be Secure.",RiskLevel::high,PolicyAction::deny);

    if (config_.require_http_only_for_sensitive_names) {
        const auto sensitive=std::any_of(config_.sensitive_name_fragments.begin(),config_.sensitive_name_fragments.end(),
            [&cookie](const std::string& fragment){ return contains_case_insensitive(cookie.name,fragment); });
        if (sensitive && !cookie.http_only) add_finding(decision,"security.sensitive.http_only","Sensitive-looking cookie name is not HttpOnly.",RiskLevel::high,PolicyAction::deny);
    }

    if (config_.require_prefix_compliance) {
        if (begins_with_case_sensitive(cookie.name,"__Secure-") && !cookie.secure)
            add_finding(decision,"prefix.secure","__Secure- cookies must be Secure.",RiskLevel::high,PolicyAction::deny);
        if (begins_with_case_sensitive(cookie.name,"__Host-")) {
            if (!cookie.secure) add_finding(decision,"prefix.host.secure","__Host- cookies must be Secure.",RiskLevel::high,PolicyAction::deny);
            if (cookie.path!="/") add_finding(decision,"prefix.host.path","__Host- cookies must use path /.",RiskLevel::high,PolicyAction::deny);
            if (!cookie.host_only || (!cookie.domain.empty() && cookie.domain.front()=='.'))
                add_finding(decision,"prefix.host.domain","__Host- cookies must be host-only.",RiskLevel::high,PolicyAction::deny);
        }
    }

    for (const auto& prefix:config_.denied_name_prefixes) {
        if (begins_with_case_sensitive(cookie.name,prefix))
            add_finding(decision,"policy.name_prefix","Cookie name uses a denied prefix: "+prefix,RiskLevel::high,PolicyAction::deny);
    }
    for (const auto& suffix:config_.denied_domain_suffixes) {
        if (domain_is_within(cookie.domain,suffix))
            add_finding(decision,"policy.domain.denied","Cookie domain is within a denied suffix: "+suffix,RiskLevel::critical,PolicyAction::deny);
    }
    if (!config_.allowed_domains.empty()) {
        const auto allowed=std::any_of(config_.allowed_domains.begin(),config_.allowed_domains.end(),
            [&cookie](const std::string& domain){ return domain_is_within(cookie.domain,domain); });
        if (!allowed) add_finding(decision,"policy.domain.allowlist","Cookie domain is not in the allowlist.",RiskLevel::critical,PolicyAction::deny);
    }
    if (config_.maximum_lifetime_days.has_value() && cookie.expiration_date.has_value()) {
        const auto lifetime_days=(*cookie.expiration_date-now_unix_seconds)/86400.0;
        if (std::isfinite(lifetime_days) && lifetime_days>*config_.maximum_lifetime_days)
            add_finding(decision,"policy.lifetime","Cookie lifetime exceeds the configured maximum.",RiskLevel::medium,PolicyAction::warn);
    }
    return decision;
}

std::vector<PolicyDecision> CookiePolicy::evaluate(const std::vector<Cookie>& cookies) const {
    std::vector<PolicyDecision> decisions;
    decisions.reserve(cookies.size());
    for (const auto& cookie:cookies) decisions.push_back(evaluate(cookie));
    return decisions;
}

std::vector<Cookie> CookiePolicy::filter_allowed(const std::vector<Cookie>& cookies) const {
    std::vector<Cookie> allowed;
    allowed.reserve(cookies.size());
    for (const auto& cookie:cookies) if (evaluate(cookie).allowed()) allowed.push_back(cookie);
    return allowed;
}

std::string risk_level_name(RiskLevel risk) {
    switch (risk) {
        case RiskLevel::information:return "information";
        case RiskLevel::low:return "low";
        case RiskLevel::medium:return "medium";
        case RiskLevel::high:return "high";
        case RiskLevel::critical:return "critical";
    }
    return "information";
}

std::string policy_action_name(PolicyAction action) {
    switch (action) {
        case PolicyAction::allow:return "allow";
        case PolicyAction::warn:return "warn";
        case PolicyAction::deny:return "deny";
    }
    return "allow";
}

RiskLevel maximum_risk(const PolicyDecision& decision) noexcept {
    RiskLevel result=RiskLevel::information;
    for (const auto& finding:decision.findings) if (finding.risk>result) result=finding.risk;
    return result;
}

}

File: rust/crates/luck-core/src/policy.rs
use crate::model::{Cookie, SameSite};
use crate::validation::{validate_cookie, IssueSeverity, ValidationConfig};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RiskLevel {
    Informational,
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PolicyViolation {
    pub rule: String,
    pub message: String,
    pub risk: RiskLevel,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum PolicyDecision {
    Allow,
    AllowWithWarnings(Vec<PolicyViolation>),
    Deny(Vec<PolicyViolation>),
}

impl PolicyDecision {
    pub fn is_allowed(&self) -> bool {
        !matches!(self, Self::Deny(_))
    }

    pub fn violations(&self) -> &[PolicyViolation] {
        match self {
            Self::Allow => &[],
            Self::AllowWithWarnings(violations) | Self::Deny(violations) => violations,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CookiePolicy {
    pub validation: ValidationConfig,
    pub require_secure: bool,
    pub require_http_only_for_names: Vec<String>,
    pub deny_name_prefixes: Vec<String>,
    pub deny_domain_suffixes: Vec<String>,
    pub allow_domains: Vec<String>,
    pub maximum_lifetime_days: Option<u64>,
    pub maximum_total_bytes: Option<usize>,
    pub deny_same_site_none: bool,
    pub deny_session_cookies: bool,
    pub deny_empty_values: bool,
}

impl Default for CookiePolicy {
    fn default() -> Self {
        Self {
            validation: ValidationConfig::default(),
            require_secure: false,
            require_http_only_for_names: vec!["session".into(), "auth".into(), "token".into()],
            deny_name_prefixes: Vec::new(),
            deny_domain_suffixes: Vec::new(),
            allow_domains: Vec::new(),
            maximum_lifetime_days: Some(400),
            maximum_total_bytes: Some(4096 + 256),
            deny_same_site_none: false,
            deny_session_cookies: false,
            deny_empty_values: false,
        }
    }
}

impl CookiePolicy {
    pub fn evaluate(&self, cookie: &Cookie) -> PolicyDecision {
        let mut violations = Vec::new();
        let validation = validate_cookie(cookie, &self.validation);
        for issue in validation.issues {
            violations.push(PolicyViolation {
                rule: issue.code,
                message: issue.message,
                risk: if issue.severity == IssueSeverity::Error { RiskLevel::High } else { RiskLevel::Low },
            });
        }
        if self.require_secure && !cookie.secure {
            violations.push(violation("policy.require_secure", "policy requires Secure", RiskLevel::High));
        }
        if self.deny_same_site_none && cookie.same_site == SameSite::None {
            violations.push(violation("policy.same_site_none", "SameSite=None is denied", RiskLevel::High));
        }
        if self.deny_session_cookies && cookie.effective_session() {
            violations.push(violation("policy.session", "session cookies are denied", RiskLevel::Medium));
        }
        if self.deny_empty_values && cookie.value.is_empty() {
            violations.push(violation("policy.empty", "empty cookie values are denied", RiskLevel::Medium));
        }
        let name_lower = cookie.name.to_ascii_lowercase();
        if self
            .require_http_only_for_names
            .iter()
            .any(|needle| name_lower.contains(&needle.to_ascii_lowercase()))
            && !cookie.http_only
        {
            violations.push(violation(
                "policy.sensitive_http_only",
                "cookie name appears sensitive but HttpOnly is disabled",
                RiskLevel::High,
            ));
        }
        for prefix in &self.deny_name_prefixes {
            if cookie.name.starts_with(prefix) {
                violations.push(violation("policy.name_prefix", format!("cookie name prefix {prefix:?} is denied"), RiskLevel::High));
            }
        }
        let domain = cookie.normalized_domain().to_ascii_lowercase();
        for suffix in &self.deny_domain_suffixes {
            let suffix = suffix.trim_start_matches('.').to_ascii_lowercase();
            if domain == suffix || domain.ends_with(&format!(".{suffix}")) {
                violations.push(violation("policy.domain_suffix", format!("domain suffix {suffix:?} is denied"), RiskLevel::Critical));
            }
        }
        if !self.allow_domains.is_empty()
            && !self.allow_domains.iter().any(|allowed| domain_matches(&domain, allowed))
        {
            violations.push(violation("policy.domain_allowlist", "domain is not in the allowlist", RiskLevel::Critical));
        }
        if let Some(maximum) = self.maximum_total_bytes {
            if cookie.estimated_bytes() > maximum {
                violations.push(violation(
                    "policy.size",
                    format!("estimated cookie size {} exceeds {maximum}", cookie.estimated_bytes()),
                    RiskLevel::High,
                ));
            }
        }
        if let (Some(days), Some(expiration)) = (self.maximum_lifetime_days, cookie.expiration_date) {
            let now = std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|duration| duration.as_secs_f64())
                .unwrap_or_default();
            if expiration > now + days as f64 * 86_400.0 {
                violations.push(violation(
                    "policy.lifetime",
                    format!("cookie lifetime exceeds {days} days"),
                    RiskLevel::Medium,
                ));
            }
        }
        let deny = violations.iter().any(|entry| entry.risk >= RiskLevel::High);
        if violations.is_empty() {
            PolicyDecision::Allow
        } else if deny {
            PolicyDecision::Deny(violations)
        } else {
            PolicyDecision::AllowWithWarnings(violations)
        }
    }
}

fn violation(rule: &'static str, message: impl Into<String>, risk: RiskLevel) -> PolicyViolation {
    PolicyViolation { rule: rule.to_owned(), message: message.into(), risk }
}

fn domain_matches(domain: &str, allowed: &str) -> bool {
    let allowed = allowed.trim_start_matches('.').to_ascii_lowercase();
    domain == allowed || domain.ends_with(&format!(".{allowed}"))
}
